File: src/main/java/net/minusmc/ravenb4/module/modules/render/NameTags.java
package raven;

import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.WorldRenderer;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.MathHelper;
import net.minecraftforge.client.event.RenderLivingEvent.Specials.Pre;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import org.lwjgl.opengl.GL11;

public class NameTags extends BaseRenderer {
    private final Option yOffset = new DoubleOption("Y-Offset", 0.0, -40.0, 40.0, 1.0);
    private final Option scale = new DoubleOption("Scale", 1.0, 0.5, 5.0, 0.1);
    private final Option autoScale = new BooleanOption("Auto-scale", false);
    private final Option drawRect = new BooleanOption("Draw Rect", true);
    private final Option showHealth = new BooleanOption("Show Health", true);
    private final Option showHitsToKill = new BooleanOption("Show Hits to Kill", false);
    private final Option showInvis = new BooleanOption("Show Invisibility", true);
    private final Option removeTags = new BooleanOption("Remove Tags", false);
    private final Option renderFriends = new BooleanOption("Render Friends (Green)", false);
    private final Option renderEnemies = new BooleanOption("Render Enemies (Red)", false);

    @SubscribeEvent
    public void renderNameTags(Pre event) {
        EntityPlayer target = event.entity;

        if (shouldRenderNametag(target)) {
            renderNametag(target, event.x, event.y, event.z);
        }
    }

    private boolean shouldRenderNametag(EntityPlayer player) {
        if (removeTags.isEnabled()) {
            return false;
        }

        if (autoScale.isEnabled()) {
            setAutoScale(player);
        } else {
            setManualScale();
        }

        return true;
    }

    private void setAutoScale(EntityPlayer player) {
        double distance = getDistanceToPlayer(player);
        double autoScaleFactor = calculateAutoScale(distance);
        double yScaleOffset = 0.0;
        setScale(autoScaleFactor, yScaleOffset);
    }

    private double getDistanceToPlayer(EntityPlayer player) {
        double xDiff = player.posX - player.prevPosX;
        double yDiff = player.posY - player.prevPosY;
        double zDiff = player.posZ - player.prevPosZ;
        return MathHelper.sqrt(xDiff * xDiff + yDiff * yDiff + zDiff * zDiff);
    }

    private double calculateAutoScale(double distance) {
        // Add your auto-scaling logic here
        double scale = 1.0; // Replace with your calculation
        return Math.max(0.5, scale);
    }

    private void setManualScale() {
        double userScale = scale.getValue();
        double yScaleOffset = yOffset.getValue();
        setScale(userScale, yScaleOffset);
    }

    private void setScale(double scaleFactor, double yScaleOffset) {
        GlStateManager.pushMatrix();
        GlStateManager.translate(0.0F, player.getEyeHeight() + 0.5F + (float) yScaleOffset, 0.0F);
        GlStateManager.scale(-scaleFactor, -scaleFactor, scaleFactor);
    }

    private void renderNametag(EntityPlayer player, double x, double y, double z) {
        String playerName = getPlayerName(player);

        if (drawRect.isEnabled()) {
            drawBackgroundRect(x, y, z);
        }

        GlStateManager.pushMatrix();
        GlStateManager.translate(x, y + player.getEyeHeight() + 0.5F, z);
        GlStateManager.rotate(-player.rotationYaw, 0.0F, 1.0F, 0.0F);

        renderNameText(playerName, x, y, z);

        GlStateManager.popMatrix();
    }

    private String getPlayerName(EntityPlayer player) {
        String playerName = player.getName();

        if (showHealth.isEnabled()) {
            playerName += " " + getFormattedHealth(player);
        }

        if (showHitsToKill.isEnabled()) {
            playerName += " " + getHitsToKill(player);
        }

        if (showInvis.isEnabled()) {
            playerName += " (Invisible)";
        }

        return playerName;
    }

    private String getFormattedHealth(EntityPlayer player) {
        return "Health: " + Math.ceil(player.getHealth());
    }

    private String getHitsToKill(EntityPlayer player) {
        // Calculate the hits to kill the player based on damage and health
        return "Hits to Kill: " + calculateHitsToKill(player);
    }

    private int calculateHitsToKill(EntityPlayer player) {
        // Implement the actual logic to calculate hits to kill
        return 0; // Replace with your calculation
    }

    private void drawBackgroundRect(double x, double y, double z) {
        Tessellator tessellator = Tessellator.getInstance();
        WorldRenderer worldRenderer = tessellator.getWorldRenderer();

        double width = 60.0;
        double height = 8.0;

        worldRenderer.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_COLOR);
        worldRenderer.pos(x - width / 2, y, z).color(0.0F, 0.0F, 0.0F, 0.7F).endVertex();
        worldRenderer.pos(x - width / 2, y + height, z).color(0.0F, 0.0F, 0.0F, 0.7F).endVertex();
        worldRenderer.pos(x + width / 2, y + height, z).color(0.0F, 0.0F, 0.0F, 0.7F).endVertex();
        worldRenderer.pos(x + width / 2, y, z).color(0.0F, 0.0F, 0.0F, 0.7F).endVertex();

        tessellator.draw();
    }

    private void renderNameText(String playerName, double x, double y, double z) {
        GlStateManager.translate(0.0F, 1.0F, 0.0F);
        GlStateManager.disableLighting();

        int textColor = 0xFFFFFF;
        b.f.drawString(playerName, -b.f.getStringWidth(playerName) / 2, 0, textColor);
        GlStateManager.enableLighting();
        GlStateManager.enableDepth();
        GlStateManager.depthMask(true);
    }
}

File: src/main/java/net/minusmc/ravenb4/module/modules/render/NameTags.kt
package net.minusmc.ravenb4.module.modules.render

import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.Tessellator
import net.minecraft.client.renderer.vertex.DefaultVertexFormats
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.MathHelper
import net.minecraftforge.client.event.RenderLivingEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minusmc.ravenb4.module.Module
import net.minusmc.ravenb4.module.ModuleCategory
import net.minusmc.ravenb4.setting.impl.SliderSetting
import net.minusmc.ravenb4.setting.impl.TickSetting
import org.lwjgl.opengl.GL11
import kotlin.math.ceil
import kotlin.math.max

class NameTags: Module("NameTags", ModuleCategory.render){

    private val yOffset: SliderSetting = SliderSetting("Y-Offset", 0.0, -40.0, 40.0, 1.0)
    private val scale: SliderSetting = SliderSetting("Scale", 1.0, 0.5, 5.0, 0.1)
    private val autoScale: TickSetting = TickSetting("Auto-scale", false)
    private val drawRect: TickSetting = TickSetting("Draw Rect", true)
    private val showHealth: TickSetting = TickSetting("Show Health", true)
    private val showHitsToKill: TickSetting = TickSetting("Show Hits to Kill", false)
    private val showInvis: TickSetting = TickSetting("Show Invisibility", true)
    private val removeTags: TickSetting = TickSetting("Remove Tags", false)
    private val renderFriends: TickSetting = TickSetting("Render Friends (Green)", false)
    private val renderEnemies: TickSetting = TickSetting("Render Enemies (Red)", false)


    init {
        addSetting(yOffset);
        addSetting(scale);
        addSetting(autoScale);
        addSetting(drawRect);
        addSetting(showHealth);
        addSetting(showHitsToKill);
        addSetting(showInvis);
        addSetting(removeTags);
        addSetting(renderFriends);
        addSetting(renderEnemies);
    }

    @SubscribeEvent
    fun renderNameTags(event: RenderLivingEvent.Specials.Pre<*>) {
        val target: EntityLivingBase = event.entity
        if(target !is EntityPlayer) return;
        if (shouldRenderNametag(target)) {
            renderNametag(target, event.x, event.y, event.z)
        }
    }

    private fun renderNametag(player: EntityPlayer, x: Double, y: Double, z: Double) {
        val playerName = getPlayerName(player)
        if (drawRect.get()) {
            drawBackgroundRect(x, y, z)
        }
        GlStateManager.pushMatrix()
        GlStateManager.translate(x, y + player.getEyeHeight() + 0.5f, z)
        GlStateManager.rotate(-player.rotationYaw, 0.0f, 1.0f, 0.0f)
        renderNameText(playerName, x, y, z)
        GlStateManager.popMatrix()
    }

    private fun getPlayerName(player: EntityPlayer): String {
        var playerName = player.name
        if (showHealth.get()) {
            playerName += " " + getFormattedHealth(player)
        }
        if (showHitsToKill.get()) {
            playerName += " " + getHitsToKill(player)
        }
        if (showInvis.get()) {
            playerName += " (Invisible)"
        }
        return playerName
    }

    private fun drawBackgroundRect(x: Double, y: Double, z: Double) {
        val tessellator = Tessellator.getInstance()
        val worldRenderer = tessellator.worldRenderer
        val width = 60.0
        val height = 8.0
        worldRenderer.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_COLOR)
        worldRenderer.pos(x - width / 2, y, z).color(0.0f, 0.0f, 0.0f, 0.7f).endVertex()
        worldRenderer.pos(x - width / 2, y + height, z).color(0.0f, 0.0f, 0.0f, 0.7f).endVertex()
        worldRenderer.pos(x + width / 2, y + height, z).color(0.0f, 0.0f, 0.0f, 0.7f).endVertex()
        worldRenderer.pos(x + width / 2, y, z).color(0.0f, 0.0f, 0.0f, 0.7f).endVertex()
        tessellator.draw()
    }

    private fun renderNameText(playerName: String, x: Double, y: Double, z: Double) {
        GlStateManager.translate(0.0f, 1.0f, 0.0f)
        GlStateManager.disableLighting()
        val textColor = 0xFFFFFF
        mc.fontRendererObj.drawString(playerName, -mc.fontRendererObj.getStringWidth(playerName) / 2, 0, textColor)
        GlStateManager.enableLighting()
        GlStateManager.enableDepth()
        GlStateManager.depthMask(true)
    }

    private fun getFormattedHealth(player: EntityPlayer): String {
        return "Health: " + ceil(player.health.toDouble())
    }

    private fun getHitsToKill(player: EntityPlayer): String {
        return "Hits to Kill: " + calculateHitsToKill(player)
    }

    private fun calculateHitsToKill(player: EntityPlayer): Int {
        var hits = (player.health / getItemDamage()).toInt();
        if(hits <= 0) hits = 0;
        return hits;
    }

    private fun getItemDamage(): Int {
        if(mc.thePlayer.heldItem == null) return 1;

        return mc.thePlayer.heldItem.itemDamage;
    }

    private fun shouldRenderNametag(player: EntityPlayer): Boolean {
        if (removeTags.get()) {
            return false
        }
        if (autoScale.get()) {
            setAutoScale(player)
        } else {
            setManualScale(player)
        }
        return true
    }

    private fun setAutoScale(player: EntityPlayer) {
        val distance = getDistanceToPlayer(player)
        val autoScaleFactor = calculateAutoScale(distance)
        val yScaleOffset = 0.0
        setScale(player, autoScaleFactor, yScaleOffset)
    }

    private fun getDistanceToPlayer(player: EntityPlayer): Double {
        val xDiff = player.posX - player.prevPosX
        val yDiff = player.posY - player.prevPosY
        val zDiff = player.posZ - player.prevPosZ
        return MathHelper.sqrt_double(xDiff * xDiff + yDiff * yDiff + zDiff * zDiff).toDouble()
    }

    private fun calculateAutoScale(distance: Double): Double {
        // Add your auto-scaling logic here
        val scale = 1.0 // Replace with your calculation
        return max(0.5, scale)
    }

    private fun setManualScale(player: EntityPlayer) {
        val userScale: Double = scale.get()
        val yScaleOffset: Double = yOffset.get()
        setScale(player, userScale, yScaleOffset)
    }

    private fun setScale(player: EntityPlayer, scaleFactor: Double, yScaleOffset: Double) {
        GlStateManager.pushMatrix()
        GlStateManager.translate(0.0f, player.getEyeHeight() + 0.5f + yScaleOffset.toFloat(), 0.0f)
        GlStateManager.scale(-scaleFactor, -scaleFactor, scaleFactor)
    }

}